src/Data/Text/IO.hs

```diff
@@ -219,42 +220,47 @@ writeLines h nl buf0 (Stream next0 s0 _len) = outer s0 buf0
           | n + 1 >= len -> commit n True{-needs flush-} False >>= outer s
           | x == '\n'    -> do
                    n' <- if nl == CRLF
-                         then do n1 <- writeCharBuf raw n '\r'
```

Comment:
I think it looked better with `raw` before, because `buf :: Buffer` contains 4 fields that are unused in these functions. It's `commitBuffer` returning a `Buffer` that's the odd one out.

I suggest to add `len` as an argument to `writeCharBuf` or to inline the bounds check in each relevant branch.

Reply:
inlined